### ISSUE TITLE
Too much fetch activity due to PeriodicWorkRequest.MIN_PERIODIC_INTERVAL and UI actions

### ISSUE THREAD

**Issue by csb0730** [deltachat/deltachat-android#1580]
- Operating System (Linux/Mac/Windows/iOS/Android): Android 4.1.2

- Delta Chat Version: 1.2.1 (but same code is existing in current sources too)

- Expected behavior:
(Imap) fetch activity should be seen every 23min when timeout of imap idle connection occurs and new idle starts again.
Except when there are new incoming messages!

- Actual behavior:
Fetch activity is started every 15min as FetchWorker.doWork() is called every 15min and interrupts all idle before any timeout occurs.
Even more worse: Every UI action triggers fetching by calling of maybe_network()
=> This generates a lot of battery drain and data traffic which is not necessary.

- Steps to reproduce the problem:
See logfile when FetchWorker.doWork() is called.

- Logs:
08-17 11:44:54.985  3594  5272 I DeltaChat: -------------------- FetchWorker.doWork() started --------------------
08-17 11:44:54.985  3594  5272 I DeltaChat: ThreadId(16) src/job.rs:667: interrupt_inbox_idle: called
...
08-17 11:59:55.090  3594  6200 I DeltaChat: -------------------- FetchWorker.doWork() started --------------------
08-17 11:59:55.090  3594  6200 I DeltaChat: ThreadId(21) src/job.rs:667: interrupt_inbox_idle: called
...
08-17 12:14:55.650  3594  5272 I DeltaChat: -------------------- FetchWorker.doWork() started --------------------
08-17 12:14:55.650  3594  5272 I DeltaChat: ThreadId(16) src/job.rs:667: interrupt_inbox_idle: called

When I increase the related time value to 30min the 15min interval is still there! Cannot explain that.

Changed source in _ApplicationContext.java_:

```
    PeriodicWorkRequest fetchWorkRequest = new PeriodicWorkRequest.Builder(
            FetchWorker.class,
            //(PeriodicWorkRequest.MIN_PERIODIC_INTERVAL_MILLIS * 2), // usually 15 minutes (cs: * 2)
            30,  //   <== even this value doesn't work !!
            TimeUnit.MINUTES,
            PeriodicWorkRequest.MIN_PERIODIC_FLEX_MILLIS, // the start may be preferred by up to 5 minutes, so we run every 10-15 minutes
            TimeUnit.MILLISECONDS)
            .setConstraints(constraints)
            .build();
```

(maybe this issue would better belong to core?)

**Comment by r10s** [comment]
the ui calls dc_maybe_network() when every 15 minutes or so, when network-connectivity changes and when the app comes to foreground, this is on-purpose.

the app is supposed to run in background all the time anyway - so, not sure, if these relatively rare events do anything on battery - do you have some more hints that this has a larger effect?

but anyway, there is not much android can do on that, might be that that can be improved, however, this is probably more on the core-side. also note, if in doubt, we vote for connectivity stability :)

**Comment by csb0730** [comment]
what me wonders is that if I increase the interval of _PeriodicWorkRequest_ from 15 min (PeriodicWorkRequest.MIN_PERIODIC_INTERVAL_MILLIS) to 30 min, I still see the 15 min calls in log!

regarding battery drain:
Since I switched from 0.500 to 1.21 I see quite double of battery drain for DC (4% ==> 8% in avarage). When I use DC more extensive it's no problem to come to 10% now!
Even UI responsiveness is feeling slightly slower than before.

In the moment I'm examining some possibilities and the extensive fetching has been coming in focus as one of the possible causes for battery drain. 
I need to say, that I have some hours every day with bad network conditions. Especially while this time battery drain rises significantly.

**Comment by csb0730** [comment]
@r10s
 are You able to confirm that when You increase interval value of _PeriodicWorkRequest_ from 15 to 30 min (for example) You always see the 15 min interval in log?

**Comment by r10s** [comment]
> are You able to confirm that when increasing interval value of PeriodicWorkRequest from 15 to 30 min (for example) You still see 15 min interval in log?

i did not tested that, however, i also do not think that increasing the PeriodicWorkRequest from 15 to 30 minutes would change things wrt system-reported battery times significantly. at least it is not measurable this way.

also these reported battery times seems to be very inaccurate.

our goal is to run every second between these timestamps anyway, however, that does not really use battery as we are just waiting for a socket.

thinking it over - so, what would be an accurate measurement for that? the system keeps a socket open anyway - how expensive is an additional socket that just waits? 0.01%? but if the system measure the uptime, divided by all apps running - that is easily 50%. also, an apps holding wakelocks consumes more battery from the viewpoint of some system than the same app doing the same things not holding a wakelock - in reality, that does not make a difference.

i am not saying, we are doing everything correctly and things cannot be improved, i am sure, it can. but the key of profiling is to find out the real bottlenecks - and the reported battery times are often not-so-helpful in this regard - but always in the scope of the user, that's right.

> Since I switched from 0.500 to 1.21 

note, that the whole app is getting much larger, for various reasons, so on old systems or on systems with few memory, that can be an issue. we are always trying to improve things there, eg. the app is getting 2mb smaller since 1.0 beginning of this year, but generally, there is no concrete, actionable item. at least i do not know of such.

> Remark: I have some hours every day with bad network conditions. Especially while this time battery drain rises significantly.

this is maybe the bad hint we have here - but on bad network conditions, increased battery usage seems to be normal, the app tries much harder to get connected, is failing, trying again etc.

**Comment by csb0730** [comment]
I'm coming back to this issue now, but I think the root cause for slow app is found ;-)

**Comment by csb0730** [comment]
@r10s 
I understand all your argumentation and are most of times with you.
But two key issues are described here and somehow mixed because especially under bad network conditions they depend somehow one from the other.

One key point of this conversation is _PeriodicWorkRequest.MIN_PERIODIC_INTERVAL_.
It triggers all 15min to interrupt all idles and leads to mailbox fetching. Due to that fact we never see a timeout of IMAP idle.
So do we use it correctly? Is it necessary all 15min? In principle IMAP idle should trigger under normal conditions.
This question has lead me to try a longer _PeriodicWorkRequest_ (PWR) interval.
=> but this doesn't work and that was strange to me. It seems to me that _PWR_ interval only triggers with 15min time span, no other! At least at my Android version (4.1.2).

Second key point is, most of user actions are triggering the same mechanism than _PWR_ (MaybeNetwork).
If you are working with DC this triggers again many fetch activities.

Third, SMTP connection. Status is checked and interrupted every minute (as I have in mind). This comes on top of the activities.

I mean, at the end the core should guarantee all these actions and should execute them as close as possible together (from timing perspective) to maximize sleep (waiting) time of the app on the other hand.

I know that _PWR_ is necessary as we don't know if the core isn't executed reliably sometimes. But the question is, is it doing the right job?
At the end it should verify if core is running and if not should restart necessary actions. This would minimize apps workload as required.

A long text so I would summarize the key questions in next separate comment.

**Comment by csb0730** [comment]
Summary of key questions:
1) is PeriodicWorkRequest.MIN_PERIODIC_INTERVAL (PWR) working korrectly? That means, is another timing than 15 min possible?
2) is _PWR_ triggiering the right actions. Means: is it doing the right job (in the moment it triggers directly attached actions; _MaybeNetwork_).
3) is it ok to trigger _MaybeNetwork_ directly by UI interactions?
4) is it ok to handle SMTP actions independently from IMAP jobs (from timing perspective)?

5) overlapping actions of core due to multiple calls of MaybeNetwork => but I think this is solved meanwhile with newer core versions (I'm working on top of 1.27). So we forget this.

IMHO all this is working, but not in an appropriate way (and leads especially under bad network conditions) to a high battery drain. That's my personal experience. 

So, meanwhile, I'm working in disentangle jobs and actions in a way I described before.

Up to now IMHO with good results.

I'm coming back with the details.

**Comment by r10s** [comment]
i think, all questions can be answered with "yes".

to

> SMTP connection. Status is checked and interrupted every minute (as I have in mind). This comes on top of the activities.

i do not think, this is true. SMTP is connected only as needed.

i am closing this issue here, i think the initial question is answered and there is no concrete actionable item left.

once there are more concrete findings we can open a more concrete issue.

**Comment by csb0730** [comment]
@r10s 
IMHO I don't think that we can answer all questions with yes simply. I stated all these questions because I found some concerns around all of them:

1. is PeriodicWorkRequest.MIN_PERIODIC_INTERVAL (PWR) working korrectly? That means, is another timing than 15 min possible?
=> at least at my device a longer interval than 15 min is not possible. And IMHO I think a longer interval is needed to get IMAP idle timeout working as required.

2. is PWR triggiering the right actions. Means: is it doing the right job (in the moment it triggers directly attached actions; MaybeNetwork).
=> MaybeNetwork should inform the core about online/offline status change of the device. In the moment it interrupts all idles without respect of online status. So it often triggers unnecessary or even wrong actions. 

3. is it ok to trigger MaybeNetwork directly by UI interactions?
=> maybe yes, but see (2). When MaybeNetwork is triggering the right actions it is ok to trigger it from UI.

4. is it ok to handle SMTP actions independently from IMAP jobs (from timing perspective)?
=> from the energy point of view, it is a difference if periodic smtp actions are done together or synced with IMAP or not.
If you have a synced operation the devices sleep time is longer, in general.

When the working principles of the core are not changed since v1.27 (except more parallel operations) then all the described concerns are still there.

IMHO the blind network operation, without knowledge if device is online and
the many many fetch operations, mainly triggered by UI concerns me.

**Comment by csb0730** [comment]
With my reworked approach I'm getting a much better result than with original one:
=> battery drain drops from 7-10% to 2-5%
Reliability of incoming messages is unchanged as far I can see with my device.

I reached this goal by
- blocking all MaybeNetwork calls <30s, even the double and parallel ones.
- introducing a real online status. When online, jobs are done, when offline jobs are not executed and idles are not interrupted.
- MaybeNetwork transfers online state from device to core again
- when MaybeNetwork is called from PeriodicWorkInterval, then call blocked when < 3600 s.
- smtp wakeup interval from 5 to 20 min
- IMAP idle now fully uses it's 23min interval from core to renew idle state.

**Comment by r10s** [comment]
some points:

- wondering why [`dc_maybe_network()`](https://c.delta.chat/classdc__context__t.html#a2f61e875270dbb7e7a2533461baa659b) takes "much time" at all. normally, this should be finished within a fraction of a second in any case. maybe this is the core of the whole thing.  
@csb0730 it this is not the case for you - can you file a dedicated issue with as many information as possible for deltachat-core-rust?  maybe it is reproducible even in the repl tool?

- introducing "online state" or things like that will cause other issues. in fact, we had that, it turns out, that this is unrealiable and led to worsening message sending/receiving. in flaky networks you just cannot know if network will work at some point - this may led to increased battery usage at a first glabce - however, messages not being sent/received also led to more battery, ppl will switch screen on more often, will double-check things etc.  
so i would not change that.

- increasing IMAP idle time to its maximum also seems reasonable, maybe do a pr on core and discuss this there.

**Comment by csb0730** [comment]
Hi @r10s 
yes, most of these issues I described here are better belonging to the core.
But maybe I did a somehow misunderstanding description at some point. As a last ;-) comment here let me clarify this:

1. dc_maybe_network() is returning quickly. I think this is a misunderstanding and that wasn't the issue I wanted to express.
It is starting it's actions in a new thread and simply interrupts the other waiting threads => ok.
The issue I detected is (I described this with < 30s) that the device often triggers dc_maybe_network() several times in short terms and then a running action is overlapped by second one. I saw this in logfile and wondered what would going on there!
In any case not ok or at least improper.

2. "online state"
yes again, "online state" from device (and now we are at Android side) is not reliable. But at the end we need to have such a status information (in core!) to control the jobs and actions appropriately.
The device's information alone is not sufficient, but should be a base. Status needs to be feed by job and connection results too. All this together gives an overall status what is more precise and really usable. I've evaluated many logs and can see this.
With this method now there is quite no job any more which fails. Even under flaky network conditions.
Maybe I post some logs to show, but not here now ;-)
In general I found the approach in trying jobs without respect to network status not suitable to handle this problem properly.

All my investigations regarding this issue here ("Too much fetch ...") started at the Android side and ended up in core!

I think the small issue around _PeriodicWorkRequest_ is not so important because anyway the core needs to control all internal activities.
But the core needs to be feed with information as accurate as possible from _PWR_, the UI actions and device status.

3. IMAP idle timeout
This simply never worked because _PWR_ triggered in 15 min intervals, which is always shorter than IMAP idle connection timeout of 23 min.  ;-)

Ok, so far.

This is the point now to switch to core. I'll open a more detailed issue about network handling approach.
This is worth to discuss because it's the base of DC's communication. Maybe some points are even covered by current core meanwhile but who knows, maybe not ;-)

**Comment by r10s** [comment]
> But at the end we need to have such a status information (in core!) to control the jobs and actions appropriately.

as said, this does not work this way. whatever status is reported at some point, it may not be the real status.

**Comment by csb0730** [comment]
I think it's possible to build a "good" status for core (meanwhile!), but let's come back to the _PeriodicWorkRequest_:

While trying and testing all last weeks finally I found this, which supports my observations:
[PeriodicWorkRequest not working correctly 69](https://github.com/googlecodelabs/android-workmanager/issues/69)
=> _PeriodicWorkRequest_ is supporting only one fixed 15 min interval. Maybe this is fixed meanwhile or depends from device, but with my device this is the fact.

With this in mind IMHO the key question of this issue here is answered now.

=> Close ok

As workaround, to get a longer _PeriodicWorkRequest interval_ I skipped every 2nd/3rd call into core. This mitigates the fixed 15 min interval and provided the longer timeout needed for more testing.

For all other questions around timeout and connection handling meanwhile I have new findings which I want to discuss in a core's issue.
